### PR TITLE
Initial import for query API

### PR DESCRIPTION
**Pull request by matthiasbeyer**
* Including Query::next() functionality
* Including some basic tests

@TheNeikos I really would like you to review this.

This is a query-API, where one can define Queries which might be executed on a TOML document. The execution routine then traverses a path (for the user) and executes the query on a sub-document of the document.

This way one can define simple conversions and updates to parts of a toml document.

---

One thing I want to do after this PR includes generalizing the _whole_ API over `Deserialize + Serialize` ... I guess this can be simplified (from a users point of view) even more with this, right?

### CODE REVIEW

File: src/query.rs [hunk 1]
```diff
@@ -0,0 +1,196 @@
...
+                        let _ = q.execute(v);
+                        next.execute(v)
+                    })
+                    .unwrap_or(Err(EK::NotAvailable(String::from(q.query_path())).into()))
+            },
+
+            None => {
+                self.read_mut(q.query_path())?
+                    .map(|v| q.execute(v))
+                    .unwrap_or(Err(EK::NotAvailable(String::from(q.query_path())).into()))
```
Comment (TheNeikos) [inline src/query.rs]: Call to `into`?

File: src/query.rs [hunk 2]
```diff
@@ -0,0 +1,196 @@
...
+    }
+
+    #[test]
+    fn test_type_converting_query_1() {
+        let mut toml : Value = toml_from_str(r#"
+        [table]
+        a = 1
+        "#).unwrap();
+
+        let query = TestQuery {};
```
Comment (TheNeikos) [inline src/query.rs]: You don't need the `{}` here

File: src/query.rs [hunk 3]
```diff
@@ -0,0 +1,196 @@
...
+
+    impl Query for PersonQuery {
+        type Output = Person;
+
+        fn execute<'doc>(&self, t: &'doc mut Value) -> Result<Self::Output> {
+            use error::ErrorKind as EK;
+
+            let first_name = t.read("first_name")?.ok_or(EK::NotAvailable("first_name".to_owned()))?;
+            let last_name = t.read("last_name")?.ok_or(EK::NotAvailable("last_name".to_owned()))?;
+            let age = t.read("age")?.ok_or(EK::NotAvailable("age".to_owned()))?;
```
Comment (TheNeikos) [inline src/query.rs]: Consider making `EK::NotAvailable` a static Cow or similar

File: src/query.rs (hunk 1 above)
Comment (matthiasbeyer) [reply src/query.rs]: Because `ErrorKind` -> `Error`.

**Comment by matthiasbeyer** [comment]
@TheNeikos Thanks for your review! What do you think about the general approach?

**Comment by TheNeikos** [comment]
Hm, from what I see right now it seems like a generalized Serde?

**Comment by matthiasbeyer** [comment]
Yes, but with full CRUD support. As far as I know, with Serde you can only do `Toml -> Type` and `Type -> TOML` conversion. With this you can do updates, for example.

Also, I'd like to generalize this (whole crate) over serialize/deserialize, but I fear that this is either out of scope or too hard.

**Pull request closed without merging**